### PR TITLE
Adding Github Oauth and group sync

### PR DESCRIPTION
**Pull request by alejandrosame**

### CODE REVIEW

File: src/website/shared/auth/github_state.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+
+# Sync functions # TODO: add function for user init next
+def sync_groups_with_github_teams() -> None:
+    """
+    Update group memberships for all users in the database based on their Github team memebership.
+    """
+    logger.info("Retrieving Github IDs to update database groups...")
+    gh_security_team_ids = get_gh_security_team_ids()
+    gh_committers_team_ids = get_gh_committers_team_ids()
```
[the remark this answers is not in the document]
Comment (alejandrosame) [reply src/website/shared/auth/github_state.py]: Done

File: src/website/shared/auth/github_state.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+) -> None:
+    """
+    Setup group memberships for a newly created user.
+    """
+    # Ignore updates and deletions
+    if not created:
+        return
+
+    logger.info("New Github account: %s. Setting up groups...", instance)
+
```
[the remark this answers is not in the document]
Comment (alejandrosame) [reply src/website/shared/auth/github_state.py]: Removed

File: src/website/shared/auth/github_state.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    if security_team.has_in_members(gh_named_user):
+        instance.user.groups.add(security_group)
+
+    if committers_team.has_in_members(gh_named_user):
+        instance.user.groups.add(committers_group)
+
+
+def sync_team_membership_from_webhook(
+    action: str, team_id: int, github_id: int
```
[the remark this answers is not in the document]
Comment (alejandrosame) [reply src/website/shared/auth/github_state.py]: Done

File: src/website/shared/auth/github_state.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+) -> None:
+    """
+    Update group membersip fro the payload received via github webhook.
```
[the remark this answers is not in the document]
Comment (alejandrosame) [reply src/website/shared/auth/github_state.py]: Fixed

File: src/website/shared/auth/github_state.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    """
+
+    user: User | None = User.objects.filter(
```
[the remark this answers is not in the document]
Comment (alejandrosame) [reply src/website/shared/auth/github_state.py]: Done

File: src/website/shared/auth/github_webhook.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+from django.http import HttpRequest, HttpResponse, HttpResponseForbidden
+from django.views.decorators.csrf import csrf_exempt
+
+from .github_state import sync_team_membership_from_webhook
+
+logger = logging.getLogger(__name__)
+
+
+def handle_webhook(event: str, payload: dict) -> None:
+    """Simple webhook handler that prints the event and payload to the console"""
```
[the remark this answers is not in the document]
Comment (alejandrosame) [reply src/website/shared/auth/github_webhook.py]: Removed

File: src/website/shared/auth/github_webhook.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        sync_team_membership_from_webhook(
+            action=payload["action"],
+            team_id=payload["team"]["id"],
+            github_id=payload["member"]["id"],
+        )
+
+
+@csrf_exempt
+def handle_github_hook(request: HttpRequest) -> HttpResponse:
+    # Check the X-Hub-Signature header to make sure this is a valid request.
```
[the remark this answers is not in the document]
Comment (alejandrosame) [reply src/website/shared/auth/github_webhook.py]: Fixed

File: src/website/shared/tests/test_github_sync.py [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+    def test_sync_groups_with_teams_is_idempotent(self) -> None:
+        # Setup mock GitHub state
+        gh_state = GithubStateMock(
+            security_ids=[self.security_users[0]["uid"]],
+            committer_ids=[self.committer_users[0]["uid"]],
+        )
+        # Sync once
+        gh_state.sync_groups_with_github_teams()
+
+        # Call sync again to check for imdempotency (no errors should be raised)
```
Comment (alejandrosame) [inline src/website/shared/tests/test_github_sync.py:229]:
```suggestion
        # Call sync again to check for idempotency (no errors should be raised)
```

File: src/website/shared/tests/test_github_sync.py [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+        # Anybody else gets read permissions
+        cls.reader_users = create_user_with_sociallogin(
+            name="reader", uid_base=30, amount=1
+        )
+        cls.user_without_social = User.objects.create_user(
+            username="user-without-social"
+        )
+
+    def test_users_without_sociallogin(self) -> None:
+        gh_state = GithubStateMock()
```
[the remark this answers is not in the document]
Comment (alejandrosame) [reply src/website/shared/tests/test_github_sync.py]: Applied as a test refactor

File: src/website/tracker/settings.py [hunk 10]
```diff
@@ -279,6 +278,21 @@ def get_secret(name: str, encoding: str = "utf-8") -> str:
     "[::1]",
 ]
 
+# Github auth settings
+GH_ORGANIZATION = "NixOS"
+GH_COMMITTERS_TEAM = "nixpkgs-committers"
+GH_SECURITY_TEAM = "security"
+GROUP_COMMITTERS = (
```
[the remark this answers is not in the document]
Comment (alejandrosame) [reply src/website/tracker/settings.py]: Applied in a separate commit

File: src/website/shared/auth/github_state.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+        if self.committers_team.id == github_team_id:
+            if action == "added":
+                user.groups.add(self.committers_group)
+            elif action == "removed":
+                user.groups.remove(self.committers_group)
+
+
+# On social sign up receiver
+@receiver(user_signed_up)
```
[the remark this answers is not in the document]
Comment (alejandrosame) [reply src/website/shared/auth/github_state.py:114]: Filed as an issue: https://github.com/Nix-Security-WG/nix-security-tracker/issues/166

**Review by fricklerhandwerk — APPROVED** [review]

**Pull request merged**